### PR TITLE
1265 lb data replay without collection

### LINKED ISSUE #1265: Implement general test driver for LB testing
**What Needs to be Done?**

Read in stats and recreate collection with input LB stats to test/debug LB strategies

### PR DESCRIPTION
**Pull request by nlslatt**
This PR allows replaying LB data (workloads) read from json files. It does not correctly handle communications and thus will not work with comm-aware load models or load balancers. To handle communications correctly, a vt refactor where communications are treated analogously to load data by the `ProposedReassignment` load model and the `LBManager` would likely be the best approach.

Closes #1265

### CODE REVIEW

**Comment by nlslatt** [comment]
This PR includes an executable `tools/workload_replay/simulate_replay` that can be used to replay an application run from stored json files but allowing new LB decisions. I will provide more details on how to use the tool later.

Due to the vt refactor that moved printing the phase summary (including LB statistics) out of the LB manager and into the phase manager, there are currently no phase summaries being printed from the replay tool (due to only simulating phases, not truly advancing through them). I'm looking into fixing that.

**Comment by nlslatt** [comment]
I added code to print the phase summary. It requires modifying private members of the LB manager. I added accessors to do this, but we should consider a refactor.

**Comment by nlslatt** [comment]
**Using the replay tool**

- First run the application as usual, dumping LB data files using the command-line option `--vt_lb_data`. For example:
```
mpirun -np 4 ./examples/collection/lb_iter 128 1 20 --vt_lb_data --vt_lb_data_dir="lbiter" --vt_lb_data_file="lbdata.%p.json.br" --vt_lb --vt_lb_name=HierarchicalLB --vt_lb_interval=5
```
where `20` is the number of phases to run.
- Then, replay from the LB data files using the replay tool with the command-line option `--vt_lb_data_in`. For example:
```
mpirun -np 4 ./tools/workload_replay/simulate_replay 0 20 --vt_lb_data_in --vt_lb_data_dir_in="lbiter" --vt_lb_data_file_in="lbdata.%p.json.br" --vt_lb --vt_lb_name="TemperedLB" --vt_lb_interval=2
```
where `0` is the initial phase to run and `20` is the number of phases to run. It is also possible to start replay well into the run, for example:
```
mpirun -np 4 ./tools/workload_replay/simulate_replay 12 8 --vt_lb_data_in --vt_lb_data_dir_in="lbiter" --vt_lb_data_file_in="lbdata.%p.json.br" --vt_lb --vt_lb_name="TemperedLB" --vt_lb_interval=2
```
which runs 8 phases, starting with phase 12.

File: src/vt/vrt/collection/balance/workload_replay.cc [hunk 1]
```diff
@@ -0,0 +1,460 @@
...
+      };
+      auto cb = theCB()->makeFunc<ReassignmentMsg>(
+        vt::pipe::LifetimeEnum::Once, postLBWork
+      );
+      theLBManager()->selectStartLB(phase, cb);
+    });
+    runInEpochCollective("WorkloadReplayDriver -> destroyLB", [&] {
+      theLBManager()->destroyLB();
+    });
+    theCollective()->barrier();
```
Comment (lifflander) [inline src/vt/vrt/collection/balance/workload_replay.cc:230]: Can we remove this barrier? I don't think it's needed

**Review by lifflander — CHANGES_REQUESTED** [review]

File: CMakeLists.txt [hunk 2]
```diff
@@ -114,6 +115,26 @@ if (VT_BUILD_TESTS
   include(CTest)
 endif()
 
+#
```
[the remark this answers is not in the document]
Comment (nlslatt) [reply CMakeLists.txt:118]: Based on my bad experiences with Trilinos options, I think it is case sensitive. If I change this, I should change the existing `VT_BUILD_TESTS` too. Are we ready to make that breaking change?

File: tools/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  if (vt_unity_build_enabled)
+    set_target_properties(${tool_name} PROPERTIES UNITY_BUILD ON)
+  endif()
+
+  link_target_with_vt(
+    TARGET ${tool_name}
+    DEFAULT_LINK_SET
+  )
+
+### @todo Add command-line arguments for testing
```
[the remark this answers is not in the document]
Comment (nlslatt) [reply tools/CMakeLists.txt:25]: Are you okay with not testing the tool itself? We have unit tests of the base functionality but not of the driver.

File: CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (nlslatt) [reply CMakeLists.txt:118]: And should I move `VT_BUILD_TESTS` to `cmake/configure_options.cmake` too?

[the remark this answers is not in the document]
Comment (lifflander) [reply CMakeLists.txt:118]: I'm think we shouldn't change `VT_BUILD_TESTS` and `VT_BUILD_EXAMPLES` right now, but should create an issue for that. Those variables were created a very long time ago, before we adopted the lower-case convention. Really I think we should have all the options in configure_options.cmake if possible. I see there are still a handful in top-level CMakeLists. If you want to move them, that would be great unless there is some reason they can't be moved.

File: tools/CMakeLists.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (lifflander) [reply tools/CMakeLists.txt:25]: If we run the tool itself (beyond the tests you wrote), we need input files and other configuration, right? I think for now the tests you have are sufficient to merge this.

[the remark this answers is not in the document]
Comment (nlslatt) [reply tools/CMakeLists.txt:25]: Removed.

File: CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (nlslatt) [reply CMakeLists.txt:118]: I'm changing it to lowercase and leaving the move to the other cmake file as part of a separate issue.

File: src/vt/vrt/collection/balance/workload_replay.cc (hunk 1 above)
Comment (nlslatt) [reply src/vt/vrt/collection/balance/workload_replay.cc:230]: Removed.

File: CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (nlslatt) [reply CMakeLists.txt:118]: The new issue is #2214.

File: src/vt/vrt/collection/balance/workload_replay.cc [hunk 4]
```diff
@@ -0,0 +1,459 @@
...
+
+std::shared_ptr<LBDataHolder>
+readInWorkloads(const std::string &filename) {
+  using util::json::Reader;
+
+  Reader r{filename};
+  auto json = r.readFile();
+  auto sd = std::make_shared<LBDataHolder>(*json);
+
+  for (auto &phase_data : sd->node_data_) {
```
Comment (lifflander) [inline src/vt/vrt/collection/balance/workload_replay.cc:243]: A minor improvement would be to conditionalize these loops on `if (theConfig()->vt_debug_replay) {`

**Review by lifflander — APPROVED** [review]

**Pull request merged**